go/apps/linux_host/linux_host.go

```diff
@@ -17,247 +17,534 @@ package main
+			stderr = nil
+		}
+		spa := &syscall.SysProcAttr{
+			Setsid: true, // Create session.
+			// Setpgid: true, // Set process group ID to new pid (SYSV setpgrp)
+			// Setctty: true, // Set controlling terminal to fd Ctty (only meaningful if Setsid is set)
+			// Noctty: true, // Detach fd 0 from controlling terminal
+			// Ctty: 0, // Controlling TTY fd (Linux only)
+		}
+		daemon := exec.Cmd{
+			Path:        path,
+			Args:        os.Args,
+			Stderr:      stderr,
+			SysProcAttr: spa,
+		}
+		err = daemon.Start()
+		failIf(err, "Can't become daemon")
+		fmt.Fprintf(noise, "Linux Tao Host running as daemon\n")
+		os.Exit(0)
+	} else {
+		fmt.Fprintf(noise, "Already a session leader?\n")
+	}
+}
+
+func startHost(domain *tao.Domain) {
+
+	if *options.Bool["daemon"] && *options.Bool["foreground"] {
+		usage("Can supply only one of -daemon and -foreground")
+	}
+	if *options.Bool["daemon"] {
+		daemonize()
+	}
+
+	cfg := configureFromFile()
+	configureFromOptions(cfg)
+	host := loadHost(domain, cfg)
+
+	sockPath := path.Join(hostPath(), "admin_socket")
+	// Make sure callers can read the admin socket directory
+	err := os.Chmod(path.Dir(sockPath), 0755)
+	failIf(err, "Can't change permissions")
+	uaddr, err := net.ResolveUnixAddr("unix", sockPath)
+	failIf(err, "Can't resolve unix socket")
+	sock, err := net.ListenUnix("unix", uaddr)
+	failIf(err, "Can't create admin socket")
+	defer sock.Close()
+	err = os.Chmod(sockPath, 0666)
+	if err != nil {
+		sock.Close()
+		fail(err, "Can't change permissions on admin socket")
+	}
+
+	go func() {
+		fmt.Fprintf(noise, "Linux Tao Service (%s) started and waiting for requests\n", host.HostName())
+		err = tao.NewLinuxHostAdminServer(host).Serve(sock)
+		fmt.Fprintf(noise, "Linux Tao Service finished\n")
+		sock.Close()
+		failIf(err, "Error serving admin requests")
+		os.Exit(0)
+	}()
+
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, os.Kill, syscall.SIGTERM)
+	<-c
+	fmt.Fprintf(noise, "Linux Tao Service shutting down\n")
+	err = shutdown()
+	if err != nil {
+		sock.Close()
+		fail(err, "Can't shut down admin socket")
 	}
 
-	glog.Flush()
+	// The above goroutine will normally end by calling os.Exit(), so we
+	// can block here indefinitely. But if we get a second kill signal,
+	// let's abort.
+	fmt.Fprintf(noise, "Waiting for shutdown....\n")
+	<-c
+	fail(nil, "Could not shut down linux_host")
+}
+
+func stopHost(domain *tao.Domain) {
+	err := shutdown()
+	if err != nil {
+		usage("Couldn't connect to linux_host: %s", err)
+	}
 }
 
-func fatalIf(err error) {
+func shutdown() error {
+	sockPath := path.Join(hostPath(), "admin_socket")
+	conn, err := net.DialUnix("unix", nil, &net.UnixAddr{sockPath, "unix"})
 	if err != nil {
-		glog.Fatal(err)
+		return err
 	}
+	defer conn.Close()
+	return tao.NewLinuxHostAdminClient(conn).Shutdown()
+}
+
+func getKey(prompt, name string) []byte {
+	if input := *options.String[name]; input != "" {
+		// TODO(kwalsh) Maybe this should go to stderr?
```

Comment:
SGTM, since logtostderr/alsologtostderr might not be set, and this warning should definitely be visible